lib/ros_message.h

```diff
@@ -24,6 +37,11 @@ class RosMessage {
   {
   }
 
+  RosMessageRawBufferData raw_data() const {
+    const std::vector<char> raw_buffer_ = *raw_buffer; 
```

Comment:
Any reason why you can't just pass around the shared_ptr?

Reply:
More of a protection thing. I was hoping this would function more as a getter, and not give the caller the ability to change the underlying buffer as far as the RosMessage sees it, which is why it's passing a copy rather than the underlying buffer itself. Could pass it around as a const shared ptr too, but I didn't even want to give the caller the option. Thoughts?

Reply:
I think passing the shared_ptr will be important to ensure they don't lose the underlying buffer. But then you could pass it as a point to a const vector to prevent them from changing the data, and get the best of both worlds!

Reply:
I can't make raw_data() const anymore since I call the hydrate function. If I pass around a const shared_ptr, will that be sufficient to protect the underlying data?

Reply:
There are a couple things here - 
First, I think there's no good reason raw_data() shouldn't be marked const. It does technically change the underlying data, but to the end user this should be invisible. It might make sense for you make this function const-qualified, by making hydrate() const-qualified as well, by marking the member variables it touches as `mutable`. This makes sense in lazily-evaluated classes like you're doing here - You should be able to run `raw_data()` on a const RosMessage and there's no good reason you shouldn't.

Second, I don't think Liam is suggesting you pass a const shared_ptr, but rather a shared_ptr to const (const shared_ptr<vector<char>> vs shared_ptr<const vector<char>>. The latter would be enough to protect the underlying data from being changed

Reply:
@liamembark I had an idea that I ran by sanjeevani and I think she's on board - 
It seems like you were hoping to avoid giving public access to the underlying buffer here, is that right?
If so, instead of providing access directly as we are in this PR, maybe the RosMessage class could have a function that returns the result of deserializing the underlying buffer? Something like

```
template<typename T>
RosMessage::toMsg() const {
  T ros_msg;
  ros::serialization::IStream s(reinterpret_cast<uint8_t*>(msg->raw_data), msg->raw_data_len);
  ros::serialization::deserialize(s, ros_msg);
  return ros_msg;
}
  ```
  
  which is what we end up doing to the underlying buffer anyway?

Reply:
Even better :)

Reply:
Although I'm not sure how comfortable I would be adding roscpp as a depedency of embag.

Reply:
Yup! I just tried creating this template function (see my comment below) but looks like we'd need to add a dependency on `roscpp` to add `ros::serialization` - is there any harm in doing that for Embag? 

Namely, to add roscpp, it looks like [this is the repo](https://github.com/nicolov/ros-bazel) to add.

Reply:
Hmm yeah that's your call liam, I could see how you'd want to avoid that

Reply:
I wonder if we can just port that one function

Reply:
I don't seem to see alternatives for roscpp in bazel, but I'm not sure I'm sold in that repo being the most well-kept / official piece of code (last update was 1 year ago)

Reply:
`deserialize` seems to [depend](http://docs.ros.org/en/diamondback/api/roscpp_serialization/html/namespaceros_1_1serialization.html#a332037f1c8b43ae8ff0b194d6042502c) on the `Serializer` class. So we'd need to bring in the entire serializer class here

Reply:
Actually, maybe not. I'm going to play around with something to see if I can just bring in the part of the code we need from Serialization

Reply:
Added just the relevant code from ros::serialization in its own file -- how does licensing work? I had to make minor changes to that code (replacing ros::Exception with just std::exception) but is basically a subset of [serialization.h](http://docs.ros.org/en/diamondback/api/roscpp_serialization/html/serialization_8h_source.html) and [serialization.cpp](http://docs.ros.org/en/diamondback/api/roscpp_serialization/html/serialization_8cpp_source.html).